### DISCUSSION TITLE
Is server-side caching possible?

### DISCUSSION THREAD

**Discussion by marcus-thorn** [vercel/next.js#40086] [Help] [answered]
Apologies, I couldn't find a thread similar to what I'm asking, although I'm sure it's been asked.  I have to questions:

1. I have a collection of pages, some using `getStaticProps` and some using `getServerSideProps`. All of these pages share the same `Layout` component and therefore the same `Header` component. 

If I have to call a series of 3 requests on the server, where all 3 requests are used to populate the `Header`, is there a recommended way to call this only one time? 

For example, if I visit 5 different pages that are all using `getServerSideProps`, currently it seems like those 3 requests need to happen each time, even though they are the same data (I understand it doesn't matter much with `getStaticProps` though).

An idea I've seen is to take those initial 3 queries, call them once, and write them to the file system (nice if the data isn't dynamic). 

2. Is it possible to cache a request response on the server for different users when using `getServerSideProps`? 

I've looked at `react-query`, but it seems like the calls need to happen regardless each time `getServerSideProps` runs, although applying a key can help provide frontend requests with already-fetched server-side data. However, I don't think it can cache the same data on the server.

Another example might be a shop with thousands of products. On each product page when you fetch the product data, it would be nice to cache the data across multiple users (Just the query, as opposed to caching the whole page via the `Cache-Control` header).

**Comment by icyJoseph** [comment, marked as answer]
The issue is that, you cannot guarantee the same JavaScript contexts between calls to these pages.

For example, when building `next build`, tries to spawn workers to render the pages, and then collects the results. As far as I understand and have tried, there's no simple way to share some object between these workers, at least not now. I tried to spawn threads or Workers, but at some point the code complexity sky rockets, and then it is simpler to use a middle layer, often an in-memory-key-value-pair kind of db, such as Redis, among others.

In some cases, as you say, writing to the file system a JSON file with the data you need and load it in your `getStaticProps` pages is a possibility we must consider.

 - Is it possible to cache a request response on the server for different users when using getServerSideProps?

Sort of... If I understand correctly you are looking at something like https://nextjs.org/docs/going-to-production#caching? 

Otherwise, while it might sound plausible to spin-up an object/map/proper-cache on the server, how will this actually work in practice is not straight-forward. 

I don't want to put a code snippet, because people might just copy paste it, but in general, a Map, could be started on the server, and you can get/set values from it on demand.

Non-exhaustive list of things to consider:

- The hosting you use might use a serverless technique to run the `GSSP`, handler, so nothing is left behind.
- If there's a single server, you need to deal with revalidation, and memory leaks
- You still hit the server anyway
- If there's multiple instances of your server, and the data updates, users might see incoherent data between reloads 

At the end of the day, having a proper in-memory database handle these things is way better. You give responsibility of the handling of data to a layer built for it, while next.js does its job of rendering pages.
